step/serialization.md

```diff
@@ -1,109 +1,553 @@
+```java
+<copy>
+import java.io.ByteArrayOutputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.net.UnixDomainSocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.channels.SocketChannel;
+
+public class ClassSerializationClient {
+
+	public static void main(String[] args) throws Exception {
+		var address = UnixDomainSocketAddress.of("mnt/server");
+		try (var clientChannel = SocketChannel.open(address)) {
+
+			ByteArrayOutputStream out = new ByteArrayOutputStream();
+			MessageClass message = new MessageClass(null);
+			ObjectOutputStream objectStream = new ObjectOutputStream(out);
+			objectStream.writeObject(message);
+			ByteBuffer buf = ByteBuffer.wrap(out.toByteArray());
+
+			clientChannel.write(buf);
+		}
+	}
+}
+</copy>
+```
 
-To create a jlink based custom Java runtime image, Helidon is using a Maven profile. To use it, simply issue the following Maven command from the project directory.
+Now run `ClassSerializationClient`:
 
-```nohighlight
+```
 <copy>
-cd ~/odl-java-hol
-mvn package -Pjlink-image -Djlink.image.defaultJvmOptions="--enable-preview"
+java ClassSerializationClient.java
 </copy>
 ```
 
-![](./images/lab11-1.png " ")
+Which will throw a NullPointerException, when in `ClassSerializationServer` `toUpperCase()` is called: 
+
+```no-highlight
+Exception in thread "main" java.lang.NullPointerException: Cannot invoke "String.toUpperCase()" because the return value of "Object.toString()" is null
+	at ClassSerializationServer.main(ClassSerializationServer.java:32)
+``` 
 
-The result are impressive as the total size (the Java runtime, the application and its dependcies) went from ~320MB to ~80MB, a ~75% gain!
+This happens because in Java when a stream is being deserialized, the constructor for the class is not called and the fields are set reflectively. So even though in the version of `MessageClass` being used by `ClassSerializationServer` it should programmatically be impossible for `message` to be null. Serialization is able to break encapsulation and provide a null value, creating a difficult to track down bug. 
 
-💡 The Helidon jlink Maven profile also creates, by default, a CDS (Class Data Sharing) archive. CDS is a JDK feature that helps reduce the startup time and memory footprints of Java applications. Check [here]((https://docs.oracle.com/en/java/javase/16/vm/class-data-sharing.html#GUID-7EAA3411-8CF0-4D19-BD05-DF5E1780AA91) from additional details.
+### Injecting Code with Serialization
+The encapsulation breaking behavior of serialization can be problemmatic, but the real security risk can come from the ability to inject code via serialization. Let's take a look at an example of a [gadget chain](https://blog.redteam-pentesting.de/2021/deserialization-gadget-chain/) attack.
```

Comment:
can be -> is problematic ;-)

Reply:
And I would stop the "how to break" section here.
"Be aware that the simple example below shows one issue with serialization but there are more, and some are quite complex. At the end, and despite being a useful mechanism, serialization is also a very weak and a known vector of attack. The next section will introduce you to remedy...."